### PR TITLE
Switch nodejs bindings to TypeScript (WIP)

### LINKED ISSUE #753: [Task]: Switch nodejs bindings to TypeScript
## Description

For NodeJS binding, we need to rewrite the bindings codebase to be in TypeScript. All tests & examples should also be ported to TypeScript. 

## Motivation

A strict typing language is always better. 

## Are you planning to do it yourself in a pull request?

Maybe.

### PR DESCRIPTION
**Pull request by oddgrd**
# Description of change
Rewrite of nodejs bindings to TypeScript, including examples and tests. 
Add missing Client methods.

## Links to any relevant issues

Fixes issue #753.

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

By making sure all examples run.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

File: bindings/nodejs/.eslintignore [hunk 1]
```diff
@@ -0,0 +1,7 @@
+scripts
+target
+**/node_modules/
+**/out/
+**/tests/
+examples/offline_signing
+**/dist
```
Comment (laumair) [inline bindings/nodejs/.eslintignore]: No new line end of file.

File: bindings/nodejs/.eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+module.exports = {
+    env: {
+        node: true,
+        browser: true,
```
Comment (laumair) [inline bindings/nodejs/.eslintrc.js]: Not sure if we need to set it true?

**Comment by laumair** [comment]
Small comments @oddgrd: Could we keep the configuration file formatting consistent? I can see that for eslint, we are using json but for prettier config we are using a yml file.

File: bindings/nodejs/examples/01_get_info.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
+// Copyright 2021-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+import { Client, initLogger } from '@iota/client';
+
+// Run with command:
+// node ./dist/01_get_info.js
```
Comment (laumair) [inline bindings/nodejs/examples/01_get_info.ts:6]: 👍

File: bindings/nodejs/examples/01_get_info.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+// In this example we will get information about the node
+async function run() {
+    initLogger({
```
Comment (laumair) [inline bindings/nodejs/examples/01_get_info.ts]: Just thinking but logger and if we can remove the duplication across examples because we define it again in every example.

File: bindings/nodejs/examples/05_get_address_balance.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+// SPDX-License-Identifier: Apache-2.0
+import { Client, initLogger } from '@iota/client';
+import 'dotenv/config';
+
+// Run with command:
+// node ./dist/05_get_address_balance.js
+
+// In this example we will get the outputs of an address that has no additional unlock
+// conditions and sum the amounts and native tokens
+async function run() {
```
Comment (laumair) [inline bindings/nodejs/examples/05_get_address_balance.ts:11]: 👍  for adding this.

File: bindings/nodejs/lib/Client.ts [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+    async getNetworkInfo() {
+        const response = await this.messageHandler.callClientMethod({
+            name: 'GetNetworkInfo',
+        });
+
+        return JSON.parse(response).payload;
+    }
+
+    // TODO: proper type for queryParameters
+    async outputIds(queryParameters: QueryParameter[]): Promise<string[]> {
```
Comment (laumair) [inline bindings/nodejs/lib/Client.ts:63]: Are these defined in [iota.js ](https://github.com/iotaledger/iota.js/tree/feat/stardust)? If yes, we should just use them.

File: bindings/nodejs/types/client.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+// Copyright 2021-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+import type { MqttBrokerOptions, Node } from './network';
+
+// TODO: update this type
+export interface ClientOptions {
```
Comment (laumair) [inline bindings/nodejs/types/client.ts]: We should update this type with the latest client options.

File: bindings/nodejs/lib/Client.ts (hunk 6 above)
Comment (oddgrd) [reply bindings/nodejs/lib/Client.ts:63]: I forgot to remove this TODO, I did define this type. As far as I can see in Iota.js it is typed as string[], and in the jsdoc params here: https://github.com/iotaledger/iota.js/blob/feat/stardust/packages/iota/src/clients/plugins/indexerPluginClient.ts#L36-L65. 

I based my type (https://github.com/oddgrd/iota.rs/blob/feat/node-ts/bindings/nodejs/types/queryParameters/index.ts) on the rust type.

File: bindings/nodejs/.eslintrc.js (hunk 2 above)
Comment (oddgrd) [reply bindings/nodejs/.eslintrc.js]: I've looked into it and I can't find a reason we need this, no.

**Comment by oddgrd** [comment]
> Small comments @oddgrd: Could we keep the configuration file formatting consistent? I can see that for eslint, we are using json but for prettier config we are using a yml file.

I've had a look at some other Iota projects and a lot use .js. It also seems good for organizing the file when it grows large. So I set both to .js.

File: src/message_interface/message_handler.rs [hunk 8]
```diff
@@ -81,6 +81,8 @@ impl ClientMessageHandler {
 
     async fn call_client_method(&self, method: &ClientMethod) -> Result<ResponseType> {
         match method {
+            // TODO: Calling finish() here restricts binding clients to one response type, unlike the
```
Comment (Thoralf-M) [inline src/message_interface/message_handler.rs]: This can be removed now, we allow to get public and internal addresses as bech32 String and the conversion can be done later to hex back, no extra method needed

File: src/node_api/indexer/query_parameters.rs [hunk 9]
```diff
@@ -46,8 +46,10 @@ impl QueryParameters {
 
 /// Query parameter for output requests
 #[derive(Debug, Clone, Serialize, Deserialize, Eq)]
+#[serde(rename_all = "camelCase")]
 pub enum QueryParameter {
     /// Bech32-encoded address that should be searched for.
+    #[serde(rename = "address")]
```
Comment (Thoralf-M) [inline src/node_api/indexer/query_parameters.rs]: Why is this needed?

Comment (oddgrd) [reply src/node_api/indexer/query_parameters.rs]: That's a mistake, I'll remove it.

File: bindings/nodejs/examples/01_get_info.ts (hunk 4 above)
Comment (oddgrd) [reply bindings/nodejs/examples/01_get_info.ts]: I've resolved this by setting a default config for initLogger.

File: bindings/nodejs/examples/offline_signing/example_prepared_transaction.json [hunk 10]
```diff
@@ -2,109 +2,103 @@
...
+                18, 177, 93, 152, 247, 36, 206, 226, 89, 230, 20, 127, 45, 93,
+                60, 98, 96, 136, 48, 16, 181, 30, 150, 142, 133, 168, 210, 52,
+                218, 52, 121, 255
             ],
+            "outputs": {
+                "inner": [
+                    {
+                        "type": "Basic",
+                        "data": {
+                            "amount": 1000000,
```
Comment (Thoralf-M) [inline bindings/nodejs/examples/offline_signing/example_prepared_transaction.json]: I believe this is outdated, should be a String in the latest version

File: bindings/nodejs/lib/constants.ts [hunk 11]
```diff
@@ -0,0 +1 @@
+export const SHIMMER_TESTNET_BECH32_HRP = 'rms';
```
Comment (Thoralf-M) [inline bindings/nodejs/lib/constants.ts:4]: Maybe also add the others and also the CoinType
```JS
export enum CoinType {
    IOTA = 4218,
    Shimmer = 4219,
}
```

Comment (Thoralf-M) [reply bindings/nodejs/lib/constants.ts:4]: Or maybe something like this should be even in @iota/types, but could be done later, fine for now to keep it here

File: bindings/nodejs/examples/offline_signing/example_prepared_transaction.json (hunk 10 above)
Comment (oddgrd) [reply bindings/nodejs/examples/offline_signing/example_prepared_transaction.json]: Correct! I'll update it.

File: bindings/nodejs/lib/constants.ts (hunk 11 above)
Comment (oddgrd) [reply bindings/nodejs/lib/constants.ts:4]: I'll add it in the meantime, and we can remove it if it gets added to @iota/types.

**Review by Tuditi — DISMISSED** [review]
LGTM! :rocket:

**Review by Thoralf-M — APPROVED** [review]

**Comment by oddgrd** [comment]
@Tuditi Hi, I dismissed your review with my review changes commit, but the PR should be good to go. Thanks!

**Review by Tuditi — APPROVED** [review]
LGTM!
